app/services/cap_quantity_and_store_changes.rb

```diff
@@ -0,0 +1,50 @@
+# frozen_string_literal: true
+
+class CapQuantityAndStoreChanges
+  def initialize(order)
+    @order = order
+    @changes = {}
+  end
+
+  def call
+    cap_insufficient_stock!
+    verify_line_items
+
+    reload_order if changes.present?
+
+    changes
+  end
+
+  private
+
+  attr_reader :order, :changes
+
+  def cap_insufficient_stock!
+    order.insufficient_stock_lines.each do |line_item|
+      changes[line_item.id] = line_item.quantity
+      line_item.cap_quantity_at_stock!
+    end
+  end
+
+  def verify_line_items
+    unavailable_stock_lines_for.each do |line_item|
+      changes[line_item.id] = changes[line_item.id] || line_item.quantity
+      line_item.update(quantity: 0)
+
+      Spree::OrderInventory.new(order).verify(line_item, order.shipment)
+    end
+  end
+
+  def reload_order
+    order.line_items.reload
+    order.update_order_fees!
+  end
+
+  def unavailable_stock_lines_for
+    order.line_items.where('variant_id NOT IN (?)', available_variants_for.select(&:id))
+  end
+
+  def available_variants_for
+    OrderCycleDistributedVariants.new(order.order_cycle, order.distributor).available_variants
```

Comment:
I wonder how expensive this will be to run for every order? And I would assume when a batch of proxy orders get processed they will usually all be for the same order cycle, right?

Maybe we could optimise it at some point :thinking:

Reply:
there's definitely room for improvement. What I like about the refactoring is that it unearthed issues that went unnoticed.